### PR TITLE
fix(runtime): some ordering issues (and start work on cache mechanism)

### PR DESCRIPTION
**Pull request by bfredl**
Rough draft! But fixes some issues with #15632

### CODE REVIEW

**Comment by clason** [comment]
Yep, that seems to fix VimTeX at least! 

(I do get a message `Cannot source a directory: "/Users/clason/.config/nvim/after/"` when triggering telescope-frecency, but I don't see any actual error.)

**Comment by bfredl** [comment]
yes, it still needs some polish :]

**Comment by lervag** [comment]
So, it seems after these recent changes, a trick I'm using in VimTeX will need to be updated for neovim. I.e., I'm using this now to check if a certain syntax file exists:

```vim
if empty(globpath(&runtimepath, l:path)) | return 0 | endif
```

Something like this should work:

```vim
let l:files = has('nvim-0.6')
      \ ? nvim_get_runtime_file(l:path,v:false)
      \ : globpath(&runtimepath, l:path)
if empty(l:files) | return 0 | endif
```

But feel free to suggest something even more appropriate.

Relevant parts in VimTeX:
* https://github.com/lervag/vimtex/blob/43ba45fe94e6b162c36f06ee80ecd8b225d7c703/autoload/vimtex/syntax/nested.vim#L30
* https://github.com/lervag/vimtex/blob/43ba45fe94e6b162c36f06ee80ecd8b225d7c703/autoload/vimtex/syntax/p/luacode.vim#L35

**Comment by bfredl** [comment]
the "Cannot source a directory" thing should be fixed.

File: src/nvim/runtime.c [hunk 1]
```diff
@@ -20,6 +21,8 @@
 # include "runtime.c.generated.h"
 #endif
 
+static bool search_path_valid = false;
+static RuntimeSearchPath runtime_search_path;
```
Comment (justinmk) [inline src/nvim/runtime.c:25]: would be nice to model this as a struct, pseudocode:

```
static rtp_searchpath = {
    RuntimeSearchPath path,
    bool valid,
}
```

in general I don't understand why we don't always do this as a habit, for all kinds of "familes" of globals.

Comment (bfredl) [reply src/nvim/runtime.c:25]: it is already scoped tho: to the file. A translation unit is already a "namespace", don't need a inner platform for that.

File: src/nvim/runtime.c [hunk 2]
```diff
@@ -227,10 +314,152 @@ int do_in_path_and_pp(char_u *path, char_u *name, int flags, DoInRuntimepathCB c
...
+
+  return search_path;
+}
+
+void invalidate_search_path(void)
+{
+  search_path_valid = false;
+}
+
+void validate_search_path(void)
```
Comment (justinmk) [inline src/nvim/runtime.c:434]: nit: would be nice to have a common prefix.

```
runtime_searchpath_validate()
runtime_searchpath_invalidate()
runtime_searchpath_build()
runtime_searchpath_push()
runtime_path_is_after()
runtime_expand_pack_entry()

File: src/nvim/runtime.c [hunk 3]
```diff
@@ -227,10 +314,152 @@ int do_in_path_and_pp(char_u *path, char_u *name, int flags, DoInRuntimepathCB c
...
+        } else {
+          xfree(after);
+        }
+      }
+      FreeWild(num_files, files);
+    }
+  }
+}
+
+static bool path_is_after(char_u *buf, size_t buflen)
```
Comment (justinmk) [inline src/nvim/runtime.c:354]:
```suggestion
/// Returns true if a directory is an |after-directory|.
static bool path_is_after(char_u *buf, size_t buflen)
```

File: src/nvim/runtime.c (hunk 2 above)
Comment (bfredl) [reply src/nvim/runtime.c:434]: I will do it for the next installment of the series (very soon)

File: src/nvim/runtime.c [hunk 4]
```diff
@@ -227,10 +314,152 @@ int do_in_path_and_pp(char_u *path, char_u *name, int flags, DoInRuntimepathCB c
   return done;
 }
 
+static void push_path(RuntimeSearchPath *search_path, char *entry, bool after)
+{
+  kv_push(*search_path, ((SearchPathItem){ entry, after }));
+}
+
+static void expand_pack_entry(RuntimeSearchPath *search_path, CharVec *after_path,
```
Comment (justinmk) [inline src/nvim/runtime.c:322]: what does this do? docstring could show a brief "input => output" example

File: src/nvim/runtime.c [hunk 5]
```diff
@@ -227,10 +314,152 @@ int do_in_path_and_pp(char_u *path, char_u *name, int flags, DoInRuntimepathCB c
...
+    kv_destroy(runtime_search_path);
+    runtime_search_path = build_runtime_search_path();
+    search_path_valid = true;
+  }
+}
+
+
+
 /// Just like do_in_path_and_pp(), using 'runtimepath' for "path".
 int do_in_runtimepath(char_u *name, int flags, DoInRuntimepathCB callback, void *cookie)
```
Comment (justinmk) [inline src/nvim/runtime.c:450]: does the docstring need an update?

File: src/nvim/runtime.c (hunk 1 above)
Comment (justinmk) [reply src/nvim/runtime.c:25]: aren't the variables related? At least a common prefix would help discover that, if not a structural grouping.

Comment (bfredl) [reply src/nvim/runtime.c:25]: originally they were but then windows CI threw a fit so I changed one to fix the problem and then they were no more.

**Pull request merged**